Add CH4 Jacobian run option to carbon simulation

lizziel opened the pull request:
### Name and Institution (Required)

Name: Lizzie Lundgren
Institution: Harvard University

### Describe the update

This PR adds the capability of building and running a KPP carbon simulation with Jacobian tracers as needed for the GCHP Integrated Methane Inversion (IMI). Changes include:

1. All code used only for Jacobian tracer runs are within ifdef JACOBIAN blocks.

1. `carbon_Funcs.F90` is simplified to remove the need for species indexes and densities passed as arguments from `carbon_gases_mod.F90`. Two new subroutines are added to the file to facilitate this: `carbon_InitCarbonKPPFuncs.F90` and `carbon_CleanupCarbonKPPFuncs.F90`. These are used for all carbon simulations, not just Jacobian tracer runs.

2. All Jacobian species are expected to have name format `CH4_jac0001`, `CH4_jac0002`, etc., with the suffix 8 characters long. The presence of `_jac` indicates Jacobian species, and the prefix indicates base species. The species database information for the base species prefix is used to fetch species database information and to set initial values, e.g. `CH4` for `CH4_jac0001`, `CH4_jac0002`, etc. Species setting`Is_Jacobian` is no longer supported in the species database yaml file.

3. The `KPP/carbon` folder contains two new files: `carbon.eqn.default` and `carbon.eqn.jacobian.5`. The former is equivalent to the standard `carbon.eqn` file used to build the mechanism. The latter may be used to rebuild the mechanism to include five Jacobian CH4 tracers. A python script called `expand_carbon_eqn.py` is included in new folder `KPP/carbon/util` to create a `carbon.eqn` file for any number of Jacobian tracers.

Note that run directory changes are not included in this PR because the IMI handles Jacobian run directory changes upon IMI set up.

This PR must be merged at the same time as two other PRs:

- https://github.com/geoschem/GCClassic/pull/106
- https://github.com/geoschem/GCHP/pull/538

### Expected changes

This is a no diff update for all simulations.

### Reference(s)

None

### Related Github Issue

None

lizziel committed:
Add chemistry mechanism for IMI with 5 Jacobian tracers (CH4)

Mechanism build using KPP 3.3.0. Equations file is the same as the in the
carbon simulation but also includes CH4_0001, CH4_0002, CH4_0003, CH4_0004,
and CH4_0005.

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Replace imi_5_Jtracers KPP mechanism with carbon_CH4Jacobian

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Update carbon_Funcs

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Remove number of jacobian tracers from carbon_gases_mod

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Set number of Jacobian tracers in carbon_Funcs.F90 by checking IDs

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Update species database for Jacobian runs

Changes include:
 - separate species name into model name and species database name
 - allow use of species database entry for other named model species
 - add Jacobian run blocks within JACOBIAN_RUN CPP blocks
 - Remove usage of Is_JacobianTracer entry in yaml file

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Update carbon/carbon_Funcs.F90 to use JACOBIAN_RUN CPP flag


Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Add JACOBIAN_RUN CPP blocks to Jacobian mechanism carbon_Funcs.F90

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Remove Jacobian run carbon mechanism file carbon_Funcs.F90

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Replace Jacobian run mechanism carbon_Funcs.F90 with symlink to carbon

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Consolidate Jacobian run mechanism into existing carbon mechanism

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Change new CPP switch from JACOBIAN_RUN to JACOBIAN; use in species_mod

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

msulprizio reviewed (commented):
Thanks @lizziel. I had a look through these changes and they look good to me.

lizziel committed:
Update warning in species database mod initialization to print root only

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Fix incorrect CPP switch name for Jacobian runs

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

1Dandan commented:
Hi @lizziel, I have tested this for GCHP and it works well when we make the modification so that the restart concentrations of species with `_jac` tag will use the initial concentrations of corresponding species without `_jac` tag. 

The scripts to create KPP equations are available in https://github.com/geoschem/integrated_methane_inversion/pull/379.

Also a reminder of add `JACOBIAN` cmake option for GCClaissc `CMakeList.txt` as well.

Specifically, the changes I made for `geos-chem/Interfaces/GCHP/Chem_GridCompMod.F90` is:
```
             ! Set spc conc to background value if rst skipped or var not there
             IF ( ( RC /= ESMF_SUCCESS .OR. RST == MAPL_RestartBootstrap .OR.   &
                  RST == MAPL_RestartSkipInitial ) .AND. (INDEX( ThisSpc%Name,'_jac' )==0 )) THEN
                DO L = 1, State_Grid%NZ
                DO J = 1, State_Grid%NY
                DO I = 1, State_Grid%NX
                   IF ( L > State_Grid%MaxChemLev .AND. &
                            ( .NOT. ThisSpc%Is_Advected ) ) THEN
                      ! For non-advected spc at L > MaxChemLev, use small number
                      State_Chm%Species(IND)%Conc(I,J,L) = 1.0E-30_FP
                   ELSE
                      ! For all other cases, use the background value in spc db
                      State_Chm%Species(IND)%Conc(I,J,L) = ThisSpc%BackgroundVV
                   ENDIF
                ENDDO
                ENDDO
                ENDDO
                IF ( MAPL_am_I_Root()) THEN
                   WRITE(*,*)  &
                        '   WARNING: using background values from species database'&
                        //' for species '//trim(ThisSpc%Name) 
                ENDIF
             ENDIF
             
#ifdef JACOBIAN
             ! Do special handling if there is _jac tagger
             ! use the non-jac species initial concentrations
             IF ( INDEX( ThisSpc%Name,'_jac' )> 0 ) THEN
                State_Chm%Species(IND)%Conc = State_Chm%Species(IND_(ThisSpc%Name(1:LEN(trim(ThisSpc%Name))-8)))%Conc
                IF ( MAPL_am_I_Root()) THEN
                   WRITE(*,*)  &
                        '   INFO: using the initial concentrations of '&
                        //ThisSpc%Name(1:LEN(trim(ThisSpc%Name))-8) &
                        //' for the Jacobian tracer '//trim(ThisSpc%Name) 
                ENDIF
             ENDIF
#endif
```

lizziel committed:
Apply changes needed for Jacobian run to new species database code in 14.7

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Generalize setting Jacobian tracer restart values for any species

Adapted from code developed by Dandan Zhang (Harvard).

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Add new carbon_Funcs.F90 subroutines to KPP stubs module for carbon

This is required to build the model with mechanisms other than carbon.

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Put GCHP-only code in JACOBIAN ifdef blocks

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Update changelog

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel committed:
Add script to create carbon.eqn with any number of Jacobian tracers

Script is located in new folder KPP/carbon/util and was provided by
Dandan Zhang (Harvard).

Signed-off-by: Lizzie Lundgren <elundgren@seas.harvard.edu>

lizziel commented:
All integration tests pass for GC-Classic and GCHP. I am currently testing in the IMI with 200 Jacobian tracers.

yantosca reviewed (approved):
Thanks @lizziel.  This looks good to merge.

lizziel commented:
I verified I can run the Jacobian tracer runs with the updates in this PR without modification. This is now good to merge.

yantosca commented:
Thanks @lizziel, I'll get this into 14.7.1 now.

yantosca commented:
I have merged locally and am running integration tests.

msulprizio reviewed (approved):
Looks good. Thanks @lizziel. I will test in the IMI once merged into GEOS-Chem.

yantosca commented:
All GEOS-Chem Classic integration tests passed:
```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

CodeDir   : 42e86f2 PR #3080 post-merge hotfix: Fix incorrect id_CO2, id_CO indices
GEOS-Chem : 0ef2d016e Hotfix: Fix typo in PR #3080 causing id_CO and id_CO2 to be swapped
HEMCO     : 9fbb78e Merge PR #355 (Update to latest GitHub Actions versions)
Cloud-J   : 1dff6fe Merge pull request #31 from geoschem/release/8.0.2
HETP      : 2a99b24 Merge pull request #2 from geoschem/bugfix/initialize_local_variables

Using 24 OpenMP threads
Number of execution tests: 38

Submitted as SLURM job: 3015072
==============================================================================

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
All GEOS-Chem Classic tests were zero-diff w/r/t `gcc.4.82cf094`, except

- gc_4x5_merra2_fullchem_APM
  - Parallel error (known issue)

yantosca commented:
All GCHP integration tests passed:
```console
==============================================================================
GCHP: Execution Test Results

CodeDir       : b6be053 PR #3080 post-merge hotfix: Fix incorrect id_CO2, id_CO indices
MAPL          : d1d528e5 Merge pull request #39 from geoschem/feature/mapl_2.55.0
GMAO_Shared   : 4ddb3ec Merge pull request #2 from geoschem/feature/mapl-upgrade
ESMA_cmake    : 01bbf2f Merge pull request #9 from geoschem/feature/turn_off_f2py_builds_in_gchp
FMS           : 259759d1 Merge pull request #3 from geoschem/feature/update_gmao_libs
FVdycoreCubed : 348620f Merge PR #11 (Scale tracers by pressure ratio in first timestep)
geos-chem     : 0ef2d016e Hotfix: Fix typo in PR #3080 causing id_CO and id_CO2 to be swapped
HEMCO         : 9fbb78e Merge PR #355 (Update to latest GitHub Actions versions)
Cloud-J       : 1dff6fe Merge pull request #31 from geoschem/release/8.0.2
HETP          : 2a99b24 Merge pull request #2 from geoschem/bugfix/initialize_local_variables
GFE           : bac529f Merge pull request #40 from Goddard-Fortran-Ecosystem/feature/mathomp4/update-gftl-1.15.2

Number of execution tests: 12

Submitted as SLURM job: 3042779
==============================================================================

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
All GCHP tests were zero-diff w/r/t `gchp.4.104279d`.
